src/librustc_typeck/check/wfcheck.rs

```diff
@@ -253,16 +253,55 @@ impl<'ccx, 'gcx> CheckTypeWellFormedVisitor<'ccx, 'gcx> {
         });
     }
 
+    fn check_auto_trait(&mut self,
+                                   trait_def_id: DefId,
+                                   span: Span)
+    {
+        let predicates = self.tcx().lookup_predicates(trait_def_id);
+
+        // If we must exclude the Self : Trait predicate contained by all
+        // traits.
+        let no_refl_predicates : Vec<_> =
+            predicates.predicates.iter().filter(|predicate| {
+                match *predicate {
+                    &ty::Predicate::Trait(ref poly_trait_ref) =>
+                        poly_trait_ref.def_id() != trait_def_id,
+                    _ => true,
+            }
+            }).collect();
+
+        let trait_def = self.tcx().lookup_trait_def(trait_def_id);
+
+        // We use an if-else here, since the generics will also trigger
+        // an extraneous error message when we find predicates like
+        // `T : Sized` for a trait like: `trait Magic<T>`.
+        if !trait_def.generics.types.get_slice(ParamSpace::TypeSpace).is_empty() {
+            error_566(self.ccx, span);
+        } else if !no_refl_predicates.is_empty() {
+            error_565(self.ccx, span);
+        }
+
+    }
+
     fn check_trait(&mut self,
                    item: &hir::Item,
                    items: &[hir::TraitItem])
     {
         let trait_def_id = self.tcx().map.local_def_id(item.id);
 
         if self.tcx().trait_has_default_impl(trait_def_id) {
+            // We want to both ensure:
+            // 1) that there are no items contained within
+            // the trait defintion
+            //
+            // 2) that the definition doesn't violate the no-super trait rule
+            // for auto traits.
+
             if !items.is_empty() {
```

Comment:
Nit: move this comment and also the `if !items.is_empty()` check into `check_auto_trait`, I think.